RaftMerge/RaftMerge.tla

```diff
@@ -175,21 +215,132 @@ ClientRequest(i, r, log) ==
+
+\* Send merge request to the leader of Region B.
+ProposeMergeRequest(i) ==
+  /\ raft[i][RegionB].is_leader
+  /\ \* This request should be sent only once.
+     Len(SelectSeq(raft[i][RegionB].logs, LAMBDA log : log.type = LogPreMerge)) = 0
+  /\ raft' = InternalRequest(
+               i, RegionB,
+               [type      |-> LogPreMerge,
+                min_index |-> 1 + Min({raft[i][RegionB].match_index[j] : j \in Store})]
+             )
+  /\ UNCHANGED <<messages, region, client_vars>>
 
 \* Return TRUE if there is a log applicable to the state machine.
+\* A log is applicable if it is committed, and the target region is not in
+\* TombStone state.
 LogAppliable(i, r) ==
-  raft[i][r].apply_index < raft[i][r].commit_index
+  /\ raft[i][r].apply_index < raft[i][r].commit_index
+  /\ region[i][r] /= RegionTombStone
 
-\* Apply Raft logs to make apply_index catch up with commit_index.
-\* This simply increases apply_index.
-ApplyLog(i, r) ==
+\* Apply LogPreMerge.
+ApplyPreMergeLog(i) ==
+  LET
+    next_index == raft[i][RegionB].apply_index + 1
+  IN
+    /\ LogAppliable(i, RegionB)
+    /\ raft[i][RegionB].logs[next_index].type = LogPreMerge
+    /\ IF raft[i][RegionA].is_leader
+       THEN
+         \* If this store is the leader of regionA, make a merge proposal, and
+         \* advance apply_index.
+         LET
+           min_index == raft[i][RegionB].logs[next_index].min_index
+           commit_index == next_index
+           fetch_logs == SubSeq(raft[i][RegionB].logs, min_index, commit_index)
+         IN
+           raft' = [InternalRequest(
+                      i, RegionA,
+                      [type         |-> LogMerge,
+                       min_index    |-> min_index,
+                       commit_index |-> commit_index,
+                       entries      |-> fetch_logs]
+                    )
+                    EXCEPT ![i][RegionB].apply_index = next_index]
+       ELSE
+         \* Otherwise, only advance apply_index.
+         raft' = [raft EXCEPT ![i][RegionB].apply_index = next_index]
+    /\ region' = [region EXCEPT ![i][RegionB] = RegionMerging]
+    /\ UNCHANGED <<messages, client_vars>>
+
+\* Apply LogMerge.
+\*
+\* This action is roughly divided into two sub-actions, and executed separately.
+\* The first step copies the logs to region B, to ensure it in sync with leader
+\* B. The second step waits until the copied logs in the first step are applied,
+\* then advances apply_index and marks this region as tombstone.
+ApplyMergeLogStep1(i) ==
+  LET
+    next_index   == raft[i][RegionA].apply_index + 1
+    min_index    == raft[i][RegionA].logs[next_index].min_index
+    commit_index == raft[i][RegionA].logs[next_index].commit_index
+    new_logs     == SubSeq(raft[i][RegionB].logs, 1, min_index - 1)
+                    \o raft[i][RegionA].logs[next_index].entries
+                    \o SubSeq(raft[i][RegionB].logs, commit_index + 1, Len(raft[i][RegionB].logs))
```

Comment:
seem we don't need to consider logs >= comit_index +1

Reply:
I think we should keep it to match the semantics of `match_index` on leader B. Otherwise there may be the case `match_index[i] > Len(logs[j])`.

Reply:
After B becomes merging state, it can't accept any log except rollback log, so here I guess we can't have the log which index >= comit_index + 1

Reply:
my fault, b can append some logs if it is in merging state. But these logs cannot contain any write propose.

Reply:
Consider this scenario, there are many stores, only one follower B is synchronized very fast with leader B, others are somehow slow. So there may be the case that the `commit_index` lags very behind, but the leader and that follower's `match_index` is fresh. If we drop the logs after commit_index, the `match_index` becomes inconsistent on leader B.